p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,200 @@
+        if (cacheConfig.isOffline()) {
+            if (!localFile.isFile()) {
+                throw new IOException("Maven is offline and the requested file does not exist locally: " + uri);
+            }
+
+            return localFile;
+        }
+
+        final FTPClient client = getClient(uri);
+        final String remotePath = uri.getPath();
+        final FTPFile remoteFile = getRemoteFile(client, remotePath);
+        final boolean isRemoteMissing = remoteFile == null;
+
+        if (localFile.isFile() && (isRemoteMissing || !mustRefresh(localFile, remoteFile))) {
+            return localFile;
+        }
+
+        if (isRemoteMissing) {
+            logger.warn("Could not find file: " + uri);
+            return null;
+        }
+
+        final File parent = FileUtils.createParentDirectories(localFile);
+        final File tempFile = File.createTempFile("download", ".tmp", parent);
+
+        try (final OutputStream os = new FileOutputStream(tempFile)) {
+            if (!client.retrieveFile(remotePath, os)) {
+                final var message = client.getReplyString();
+                throw new IOException(String.format("Error retrieving file: %s. Message: %s", remotePath, message));
+            }
+        } catch (final IOException e) {
+            tempFile.delete();
+            throw e;
+        }
+
+        if (localFile.isFile()) {
+            FileUtils.forceDelete(localFile);
+        }
+
+        FileUtils.moveFile(tempFile, localFile);
+        localFile.setLastModified(remoteFile.getTimestampInstant().toEpochMilli());
+        return localFile;
+    }
+
+    public void dispose() {
+        for (final Map.Entry<String, FTPClient> entry : clients.entrySet()) {
+            try {
+                entry.getValue().logout();
+                entry.getValue().disconnect();
+            } catch (final FTPConnectionClosedException e) {
+                // Connection already closed by the host
+            } catch (final IOException e) {
+                logger.error("Error disconnecting from host: " + entry.getKey(), e);
+            }
+        }
+    }
+
+    protected FTPClient createFtpClient() {
+        final FTPClient client = new FTPClient();
+        client.setControlKeepAliveTimeout(Duration.ofSeconds(15));
+        client.setAutodetectUTF8(true);
+        client.setBufferSize(8192);
+        return client;
+    }
+
+    private FTPClient getClient(final URI uri) throws IOException {
+        final String host = uri.getHost();
+        final int port = uri.getPort() < 0 ? 21 : uri.getPort();
+        final String key = host + ":" + port;
+        final FTPClient client = clients.computeIfAbsent(key, k -> createFtpClient());
+
+        if (client.isConnected()) {
+            return client;
+        }
+
+        client.connect(host, port);
+
+        if (!FTPReply.isPositiveCompletion(client.getReplyCode())) {
+            client.disconnect();
+            throw new IOException("Could not connect to host: " + key);
```

Comment:
It might be a good ide to give the user as much information as possible, e.g the replycode + reply message if possible.

Reply:
Updated with the returned error message (includes the error code).